cpu/stm32l0/periph/adc.c

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (C) 2014-2017 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_stm32l0
+ * @ingroup     drivers_periph_adc
+ * @{
+ *
+ * @file
+ * @brief       Low-level ADC driver implementation
+ *
+ * @author      Aurélien Fillau <aurelien.fillau@we-sens.com>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "mutex.h"
+#include "periph/adc.h"
+
+#ifdef ADC_CONFIG
+
+/**
+ * @brief   Maximum allowed ADC clock speed
+ */
+#define MAX_ADC_SPEED           (12000000U)
+
+/**
+ * @brief   Load the ADC configuration
+ */
+static const adc_conf_t adc_config[] = ADC_CONFIG;
+
+/**
+ * @brief   Allocate locks for all three available ADC device
+ *
+ * All STM32L0 CPUs we support so far only come with a single ADC device.
+ */
+static mutex_t lock = MUTEX_INIT;
+
+static inline void prep(void)
+{
+    mutex_lock(&lock);
+    periph_clk_en(APB2, RCC_APB2ENR_ADCEN);
+}
+
+static inline void done(void)
+{
+    periph_clk_dis(APB2, RCC_APB2ENR_ADCEN);
+    mutex_unlock(&lock);
+}
+
+static void _enable_adc(void)
+{
+	if ((ADC1->CR & ADC_CR_ADEN) != 0) {
+		ADC1->CR |= ADC_CR_ADDIS;
+		while(ADC1->CR & ADC_CR_ADEN) {} /* Wait for ADC disabled */
+	}
+
+	if ((ADC1->CR & ADC_CR_ADEN) == 0) {
+		/* Then, start a calibration */
+		ADC1->CR |= ADC_CR_ADCAL;
+		while(ADC1->CR & ADC_CR_ADCAL) {} /* Wait for the end of calibration */
+	}
+
+	/* Clear flag */
+	ADC1->ISR |= ADC_ISR_ADRDY;
+
+	/* enable device */
+	ADC1->CR = ADC_CR_ADVREGEN | ADC_CR_ADEN;
+
+	/* Wait for ADC to be ready */
+	while (!(ADC1->ISR & ADC_ISR_ADRDY)) {}
+}
+
+static void _disable_adc(void)
+{
+    /* Disable ADC */
+    if ((ADC1->CR & ADC_CR_ADEN) != 0) {
+        ADC1->CR |= ADC_CR_ADDIS;
+        while(ADC1->CR & ADC_CR_ADEN) {} /* Wait for ADC disabled */
+        /* Disable Voltage regulator */
+        ADC1->CR = 0;
+        ADC1->ISR = 0;
+    }
+}
+
+int adc_init(adc_t line)
+{
+    /* make sure the given line is valid */
+    if (line >= ADC_NUMOF) {
+        return -1;
+    }
+
+    /* lock and power on the device */
+    prep();
+
+    if ((adc_config[line].chan != 17) && (adc_config[line].chan != 18)) {
+        /*configure the pin */
+        gpio_init_analog(adc_config[line].pin);
+    }
+
+    ADC1->CFGR1 = 0; //no watchdog, no discontinuous mode, no auto off, single conv, no trigger, right align, 12bits, no dma, no wait
```

Comment:
Please use C-style comments, e.g `/*  */` here and below. For readability and reviewability in Github, you should also split the long lines.

Reply:
done